Internal IDP Rule errors

3 Rule errors for you :)

One:
```
[null] ERROR: java.lang.NullPointerException at java.base/java.util.Objects.requireNonNull(Objects.java:208) 
at java.base/java.util.Optional.of(Optional.java:113) 
at stroom.security.impl.BasicUserIdentity.getFullName(BasicUserIdentity.java:54) 
at stroom.util.NullSafe.consume(NullSafe.java:660) 
at stroom.query.common.v2.ResultStoreManager.addCurrentUserParam(ResultStoreManager.java:265) 
at stroom.query.common.v2.ResultStoreManager.getResultStore(ResultStoreManager.java:226) 
at stroom.analytics.impl.ScheduledQueryAnalyticExecutor.processScheduledQueryAnalytic(ScheduledQueryAnalyticExecutor.java:267) 
at stroom.analytics.impl.ScheduledQueryAnalyticExecutor.lambda$processScheduledQueryAnalytic$4(ScheduledQueryAnalyticExecutor.java:202) 
at stroom.analytics.impl.AnalyticErrorWritingExecutor.lambda$wrap$0(AnalyticErrorWritingExecutor.java:37) 
at stroom.analytics.impl.AnalyticErrorWriter.exec(AnalyticErrorWriter.java:64) 
at stroom.analytics.impl.AnalyticErrorWritingExecutor.lambda$wrap$1(AnalyticErrorWriting...
```

Two:
```
[AnalyticsExecutor] FATAL: Cannot invoke "java.lang.Long.longValue()" because the return value of "stroom.query.common.v2.CurrentDbState.getLastEventTime()" is null
```


Three:
```
RACE  [2023-09-20T09:06:53.322Z] [Stroom P2 #16] stroom.pipeline.errorhandler.ErrorReceiverProxy -  java.lang.ArrayIndexOutOfBoundsException
stroom.search.extraction.ExtractionException: java.lang.ArrayIndexOutOfBoundsException
        at stroom.search.extraction.ExtractionException.wrap(ExtractionException.java:37)
        at stroom.analytics.impl.AnalyticsStreamProcessor.extract(AnalyticsStreamProcessor.java:185)
        at stroom.analytics.impl.AnalyticsStreamProcessor.processData(AnalyticsStreamProcessor.java:135)
        at stroom.analytics.impl.AnalyticsStreamProcessor.extract(AnalyticsStreamProcessor.java:112)
        at stroom.analytics.impl.StreamingAnalyticExecutor.lambda$processStream$16(StreamingAnalyticExecutor.java:356)
        at stroom.analytics.impl.AnalyticErrorWritingExecutor.lambda$wrap$0(AnalyticErrorWritingExecutor.java:37)
        at stroom.analytics.impl.AnalyticErrorWriter.exec(AnalyticErrorWriter.java:64)
        at stroom.analytics.impl.AnalyticErrorWritingExecutor.lambda$wrap$1(AnalyticErrorWritingExecutor.java:34)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$createFromConsumer$0(TaskContextFactoryImpl.java:193)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$2(TaskContextFactoryImpl.java:266)
        at stroom.util.logging.LocationAwareLambdaLogger.logDurationIfDebugEnabled(LocationAwareLambdaLogger.java:149)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$4(TaskContextFactoryImpl.java:266)
        at stroom.util.pipeline.scope.PipelineScopeRunnable.scopeResult(PipelineScopeRunnable.java:38)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$5(TaskContextFactoryImpl.java:263)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$7(TaskContextFactoryImpl.java:277)
        at stroom.security.impl.SecurityContextImpl.asUserResult(SecurityContextImpl.java:285)
        at jdk.internal.reflect.GeneratedMethodAccessor134.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at com.google.inject.internal.DelegatingInvocationHandler.invoke(DelegatingInvocationHandler.java:50)
        at jdk.proxy2/jdk.proxy2.$Proxy116.asUserResult(Unknown Source)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$11(TaskContextFactoryImpl.java:273)
        at stroom.analytics.impl.StreamingAnalyticExecutor.processStream(StreamingAnalyticExecutor.java:374)
        at stroom.analytics.impl.StreamingAnalyticExecutor.processPipeline(StreamingAnalyticExecutor.java:229)
        at stroom.analytics.impl.StreamingAnalyticExecutor.lambda$processAll$8(StreamingAnalyticExecutor.java:193)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$createFromConsumer$0(TaskContextFactoryImpl.java:193)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$2(TaskContextFactoryImpl.java:266)
        at stroom.util.logging.LocationAwareLambdaLogger.logDurationIfDebugEnabled(LocationAwareLambdaLogger.java:149)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$4(TaskContextFactoryImpl.java:266)
        at stroom.util.pipeline.scope.PipelineScopeRunnable.scopeResult(PipelineScopeRunnable.java:38)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$5(TaskContextFactoryImpl.java:263)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$7(TaskContextFactoryImpl.java:277)
        at stroom.security.impl.SecurityContextImpl.asUserResult(SecurityContextImpl.java:285)
        at jdk.internal.reflect.GeneratedMethodAccessor134.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at com.google.inject.internal.DelegatingInvocationHandler.invoke(DelegatingInvocationHandler.java:50)
        at jdk.proxy2/jdk.proxy2.$Proxy116.asUserResult(Unknown Source)
        at stroom.task.impl.TaskContextFactoryImpl.lambda$wrap$11(TaskContextFactoryImpl.java:273)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: stroom.pipeline.errorhandler.LoggedException: java.lang.ArrayIndexOutOfBoundsException
        at stroom.pipeline.errorhandler.LoggedException.wrap(LoggedException.java:44)
        at stroom.pipeline.filter.XsltFilter.endDocument(XsltFilter.java:306)
        at stroom.pipeline.filter.AbstractXMLFilter.endDocument(AbstractXMLFilter.java:158)
        at stroom.pipeline.filter.AbstractXMLFilter.endDocument(AbstractXMLFilter.java:158)
        at stroom.pipeline.filter.SplitFilter.fireRemainingEvents(SplitFilter.java:441)
        at stroom.pipeline.filter.SplitFilter.endDocument(SplitFilter.java:200)
        at org.apache.xerces.parsers.AbstractSAXParser.endDocument(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentScannerImpl.endEntity(Unknown Source)
        at org.apache.xerces.impl.XMLEntityManager.endEntity(Unknown Source)
        at org.apache.xerces.impl.XMLEntityScanner.load(Unknown Source)
        at org.apache.xerces.impl.XML11EntityScanner.skipSpaces(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentScannerImpl$TrailingMiscDispatcher.dispatch(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
        at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
        at org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source)
        at stroom.pipeline.parser.AbstractParser.parse(AbstractParser.java:184)
        at stroom.pipeline.parser.AbstractParser.lambda$createProcessors$0(AbstractParser.java:85)
        at stroom.pipeline.factory.PipelineImpl.process(PipelineImpl.java:129)
        at stroom.analytics.impl.AnalyticsStreamProcessor.lambda$extract$2(AnalyticsStreamProcessor.java:177)
        at stroom.util.logging.LocationAwareLambdaLogger.logDurationIfDebugEnabled(LocationAwareLambdaLogger.java:220)
        at stroom.analytics.impl.AnalyticsStreamProcessor.extract(AnalyticsStreamProcessor.java:176)
        ... 41 common frames omitted
```

--- a/stroom-analytics/stroom-analytics-impl/src/main/java/stroom/analytics/impl/TableBuilderAnalyticExecutor.java
+++ b/stroom-analytics/stroom-analytics-impl/src/main/java/stroom/analytics/impl/TableBuilderAnalyticExecutor.java
@@ -69,6 +69,7 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.UUID;
@@ -555,20 +556,29 @@ public class TableBuilderAnalyticExecutor {
                                  final DetectionConsumer detectionConsumer,
                                  final AnalyticDataStore dataStore,
                                  final CurrentDbState currentDbState) {
-        SimpleDuration timeToWaitForData = analytic.analyticProcessConfig.getTimeToWaitForData();
-        if (timeToWaitForData == null) {
-            timeToWaitForData = SimpleDuration.builder().time(1).timeUnit(TimeUnit.HOURS).build();
-        }
 
-        Instant from = Instant.ofEpochMilli(0);
+        final SimpleDuration timeToWaitForData = Objects.requireNonNullElseGet(
+                analytic.analyticProcessConfig.getTimeToWaitForData(),
+                () -> SimpleDuration.builder()
+                        .time(1)
+                        .timeUnit(TimeUnit.HOURS)
+                        .build());
+
+        final Instant from;
         if (analytic.trackerData.getLastWindowEndTimeMs() != null) {
             from = Instant.ofEpochMilli(analytic.trackerData.getLastWindowEndTimeMs() + 1);
         } else if (analytic.analyticProcessConfig.getMinMetaCreateTimeMs() != null) {
             from = Instant.ofEpochMilli(analytic.analyticProcessConfig.getMinMetaCreateTimeMs());
+        } else {
+            from = Instant.ofEpochMilli(0);
         }
 
-        Instant to = Instant.ofEpochMilli(currentDbState.getLastEventTime());
-        to = SimpleDurationUtil.minus(to, timeToWaitForData);
+        // TODO: 20/09/2023 Not sure what the default value should be if getLastEventTime is null
+        Instant to = Instant.ofEpochMilli(Objects.requireNonNullElse(currentDbState.getLastEventTime(), 0L));
+        final Instant newTo = SimpleDurationUtil.minus(to, timeToWaitForData);
+        if (!newTo.isBefore(Instant.EPOCH)) {
+            to = newTo;
+        }
         if (analytic.analyticProcessConfig.getMaxMetaCreateTimeMs() != null) {
             Instant max = Instant.ofEpochMilli(analytic.analyticProcessConfig.getMaxMetaCreateTimeMs());
             if (max.isBefore(to)) {

--- a/stroom-query/stroom-query-common/src/main/java/stroom/query/common/v2/CurrentDbStateFactory.java
+++ b/stroom-query/stroom-query-common/src/main/java/stroom/query/common/v2/CurrentDbStateFactory.java
@@ -3,8 +3,9 @@ package stroom.query.common.v2;
 import stroom.dashboard.expression.v1.FieldIndex;
 import stroom.dashboard.expression.v1.Val;
 import stroom.query.api.v2.SearchRequestSource.SourceType;
+import stroom.util.shared.GwtNullSafe;
 
-import java.util.Optional;
+import java.util.Objects;
 
 public class CurrentDbStateFactory {
 
@@ -42,13 +43,13 @@ public class CurrentDbStateFactory {
                     long time = 0;
                     if (timeFieldIndex >= 0 && timeFieldIndex < values.length) {
                         final Val eventTime = values[timeFieldIndex];
-                        time = Optional.ofNullable(eventTime).map(Val::toLong).orElse(0L);
+                        time = GwtNullSafe.getOrElse(eventTime, Val::toLong, 0L);
                     }
 
-                    final long streamIdLong = Optional.ofNullable(streamId.toLong()).orElseThrow(() ->
-                            new RuntimeException("Unable to get stream id"));
-                    final long eventIdLong = Optional.ofNullable(eventId.toLong()).orElseThrow(() ->
-                            new RuntimeException("Unable to get event id"));
+                    final long streamIdLong = Objects.requireNonNull(
+                            streamId.toLong(), "Unable to get stream id");
+                    final long eventIdLong = Objects.requireNonNull(
+                            eventId.toLong(), "Unable to get event id");
 
                     return new CurrentDbState(streamIdLong, eventIdLong, time);
                 }

--- a/stroom-query/stroom-query-common/src/main/java/stroom/query/common/v2/LmdbDataStore.java
+++ b/stroom-query/stroom-query-common/src/main/java/stroom/query/common/v2/LmdbDataStore.java
@@ -45,6 +45,7 @@ import stroom.util.io.FileUtil;
 import stroom.util.logging.LambdaLogger;
 import stroom.util.logging.LambdaLoggerFactory;
 import stroom.util.logging.Metrics;
+import stroom.util.shared.GwtNullSafe;
 import stroom.util.shared.time.SimpleDuration;
 
 import com.esotericsoftware.kryo.io.Input;
@@ -770,11 +771,7 @@ public class LmdbDataStore implements DataStore {
     }
 
     private void putLong(final ByteBuffer valueBuffer, final Long l) {
-        if (l == null) {
-            valueBuffer.putLong(-1);
-        } else {
-            valueBuffer.putLong(l);
-        }
+        valueBuffer.putLong(GwtNullSafe.requireNonNullElse(l, -1L));
     }
 
     public CurrentDbState sync() {

--- a/stroom-security/stroom-security-impl/src/main/java/stroom/security/impl/BasicUserIdentity.java
+++ b/stroom-security/stroom-security-impl/src/main/java/stroom/security/impl/BasicUserIdentity.java
@@ -51,7 +51,7 @@ public class BasicUserIdentity implements UserIdentity, HasStroomUserIdentity {
 
     @Override
     public Optional<String> getFullName() {
-        return Optional.of(fullName);
+        return Optional.ofNullable(fullName);
     }
 
     @Override

--- a/stroom-test-common/src/main/java/stroom/test/common/DynamicTestBuilder.java
+++ b/stroom-test-common/src/main/java/stroom/test/common/DynamicTestBuilder.java
@@ -304,8 +304,10 @@ class DynamicTestBuilder {
                         Assertions.assertThat(actualThrowable)
                                 .isInstanceOf(expectedThrowableType);
                     } else {
-                        LOGGER.debug("Test did not throw an exception but we were expecting it to throw: {}",
-                                expectedThrowableType.getSimpleName());
+                        LOGGER.debug("Test did not throw an exception but we were expecting it to throw: {}. " +
+                                        "Actual output: '{}'",
+                                expectedThrowableType.getSimpleName(),
+                                testOutcome.getActualOutput());
                         Assertions.fail("Expecting test to throw "
                                 + expectedThrowableType.getSimpleName());
                     }

--- a/stroom-util/src/main/java/stroom/util/time/SimpleDurationUtil.java
+++ b/stroom-util/src/main/java/stroom/util/time/SimpleDurationUtil.java
@@ -25,39 +25,18 @@ public class SimpleDurationUtil {
                 final String number = matcher.group(1);
                 final String unit = matcher.group(2);
                 final long time = Long.parseLong(number);
-                final TimeUnit timeUnit;
-
-                switch (unit) {
-                    case "ns":
-                        timeUnit = TimeUnit.NANOSECONDS;
-                        break;
-                    case "ms":
-                        timeUnit = TimeUnit.MILLISECONDS;
-                        break;
-                    case "s":
-                        timeUnit = TimeUnit.SECONDS;
-                        break;
-                    case "m":
-                        timeUnit = TimeUnit.MINUTES;
-                        break;
-                    case "h":
-                        timeUnit = TimeUnit.HOURS;
-                        break;
-                    case "d", "D":
-                        timeUnit = TimeUnit.DAYS;
-                        break;
-                    case "w", "W":
-                        timeUnit = TimeUnit.WEEKS;
-                        break;
-                    case "M":
-                        timeUnit = TimeUnit.MONTHS;
-                        break;
-                    case "y", "Y":
-                        timeUnit = TimeUnit.YEARS;
-                        break;
-                    default:
-                        throw new ParseException("Time unit is invalid: " + trimmed, 0);
-                }
+                final TimeUnit timeUnit = switch (unit) {
+                    case "ns" -> TimeUnit.NANOSECONDS;
+                    case "ms" -> TimeUnit.MILLISECONDS;
+                    case "s" -> TimeUnit.SECONDS;
+                    case "m" -> TimeUnit.MINUTES;
+                    case "h" -> TimeUnit.HOURS;
+                    case "d", "D" -> TimeUnit.DAYS;
+                    case "w", "W" -> TimeUnit.WEEKS;
+                    case "M" -> TimeUnit.MONTHS;
+                    case "y", "Y" -> TimeUnit.YEARS;
+                    default -> throw new ParseException("Time unit is invalid: " + trimmed, 0);
+                };
 
                 return new SimpleDuration(time, timeUnit);
             } catch (final RuntimeException e) {
@@ -80,27 +59,17 @@ public class SimpleDurationUtil {
         if (simpleDuration == null) {
             return dateTime;
         }
-        switch (simpleDuration.getTimeUnit()) {
-            case NANOSECONDS:
-                return dateTime.plusNanos(simpleDuration.getTime());
-            case MILLISECONDS:
-                return dateTime.plusNanos(simpleDuration.getTime() * 1000000);
-            case SECONDS:
-                return dateTime.plusSeconds(simpleDuration.getTime());
-            case MINUTES:
-                return dateTime.plusMinutes(simpleDuration.getTime());
-            case HOURS:
-                return dateTime.plusHours(simpleDuration.getTime());
-            case DAYS:
-                return dateTime.plusDays(simpleDuration.getTime());
-            case WEEKS:
-                return dateTime.plusWeeks(simpleDuration.getTime());
-            case MONTHS:
-                return dateTime.plusMonths(simpleDuration.getTime());
-            case YEARS:
-                return dateTime.plusYears(simpleDuration.getTime());
-        }
-        throw new UnsupportedOperationException("Unknown time unit");
+        return switch (simpleDuration.getTimeUnit()) {
+            case NANOSECONDS -> dateTime.plusNanos(simpleDuration.getTime());
+            case MILLISECONDS -> dateTime.plusNanos(simpleDuration.getTime() * 1_000_000);
+            case SECONDS -> dateTime.plusSeconds(simpleDuration.getTime());
+            case MINUTES -> dateTime.plusMinutes(simpleDuration.getTime());
+            case HOURS -> dateTime.plusHours(simpleDuration.getTime());
+            case DAYS -> dateTime.plusDays(simpleDuration.getTime());
+            case WEEKS -> dateTime.plusWeeks(simpleDuration.getTime());
+            case MONTHS -> dateTime.plusMonths(simpleDuration.getTime());
+            case YEARS -> dateTime.plusYears(simpleDuration.getTime());
+        };
     }
 
     public static Instant minus(final Instant dateTime,
@@ -115,91 +84,72 @@ public class SimpleDurationUtil {
         if (simpleDuration == null) {
             return dateTime;
         }
-        switch (simpleDuration.getTimeUnit()) {
-            case NANOSECONDS:
-                return dateTime.minusNanos(simpleDuration.getTime());
-            case MILLISECONDS:
-                return dateTime.minusNanos(simpleDuration.getTime() * 1000000);
-            case SECONDS:
-                return dateTime.minusSeconds(simpleDuration.getTime());
-            case MINUTES:
-                return dateTime.minusMinutes(simpleDuration.getTime());
-            case HOURS:
-                return dateTime.minusHours(simpleDuration.getTime());
-            case DAYS:
-                return dateTime.minusDays(simpleDuration.getTime());
-            case WEEKS:
-                return dateTime.minusWeeks(simpleDuration.getTime());
-            case MONTHS:
-                return dateTime.minusMonths(simpleDuration.getTime());
-            case YEARS:
-                return dateTime.minusYears(simpleDuration.getTime());
-        }
-        throw new UnsupportedOperationException("Unknown time unit");
+        return switch (simpleDuration.getTimeUnit()) {
+            case NANOSECONDS -> dateTime.minusNanos(simpleDuration.getTime());
+            case MILLISECONDS -> dateTime.minusNanos(simpleDuration.getTime() * 1_000_000);
+            case SECONDS -> dateTime.minusSeconds(simpleDuration.getTime());
+            case MINUTES -> dateTime.minusMinutes(simpleDuration.getTime());
+            case HOURS -> dateTime.minusHours(simpleDuration.getTime());
+            case DAYS -> dateTime.minusDays(simpleDuration.getTime());
+            case WEEKS -> dateTime.minusWeeks(simpleDuration.getTime());
+            case MONTHS -> dateTime.minusMonths(simpleDuration.getTime());
+            case YEARS -> dateTime.minusYears(simpleDuration.getTime());
+        };
     }
 
     public static LocalDateTime roundDown(LocalDateTime dateTime, SimpleDuration simpleDuration) {
-        switch (simpleDuration.getTimeUnit()) {
-            case NANOSECONDS:
-                return LocalDateTime.of(
-                        dateTime.getYear(),
-                        dateTime.getMonthValue(),
-                        dateTime.getDayOfMonth(),
-                        dateTime.getHour(),
-                        dateTime.getMinute(),
-                        dateTime.getSecond(),
-                        (dateTime.getNano() / 1000000) & 1000000);
-            case MILLISECONDS:
-                return LocalDateTime.of(
-                        dateTime.getYear(),
-                        dateTime.getMonthValue(),
-                        dateTime.getDayOfMonth(),
-                        dateTime.getHour(),
-                        dateTime.getMinute(),
-                        dateTime.getSecond());
-            case SECONDS:
-                return LocalDateTime.of(
-                        dateTime.getYear(),
-                        dateTime.getMonthValue(),
-                        dateTime.getDayOfMonth(),
-                        dateTime.getHour(),
-                        dateTime.getMinute());
-            case MINUTES:
-                return LocalDateTime.of(
-                        dateTime.getYear(),
-                        dateTime.getMonthValue(),
-                        dateTime.getDayOfMonth(),
-                        dateTime.getHour(),
-                        0);
-            case HOURS:
-                return LocalDateTime.of(
-                        dateTime.getYear(),
-                        dateTime.getMonthValue(),
-                        dateTime.getDayOfMonth(),
-                        0,
-                        0);
-            case DAYS:
-                return LocalDateTime.of(
-                        dateTime.getYear(),
-                        dateTime.getMonthValue(),
-                        1,
-                        0,
-                        0);
-            case WEEKS, MONTHS:
-                return LocalDateTime.of(
-                        dateTime.getYear(),
-                        1,
-                        1,
-                        0,
-                        0);
-            case YEARS:
-                return LocalDateTime.of(
-                        1970,
-                        1,
-                        1,
-                        0,
-                        0);
-        }
-        throw new UnsupportedOperationException("Unknown time unit");
+        return switch (simpleDuration.getTimeUnit()) {
+            case NANOSECONDS -> LocalDateTime.of(
+                    dateTime.getYear(),
+                    dateTime.getMonthValue(),
+                    dateTime.getDayOfMonth(),
+                    dateTime.getHour(),
+                    dateTime.getMinute(),
+                    dateTime.getSecond(),
+                    (dateTime.getNano() / 1_000_000) & 1_000_000);
+            case MILLISECONDS -> LocalDateTime.of(
+                    dateTime.getYear(),
+                    dateTime.getMonthValue(),
+                    dateTime.getDayOfMonth(),
+                    dateTime.getHour(),
+                    dateTime.getMinute(),
+                    dateTime.getSecond());
+            case SECONDS -> LocalDateTime.of(
+                    dateTime.getYear(),
+                    dateTime.getMonthValue(),
+                    dateTime.getDayOfMonth(),
+                    dateTime.getHour(),
+                    dateTime.getMinute());
+            case MINUTES -> LocalDateTime.of(
+                    dateTime.getYear(),
+                    dateTime.getMonthValue(),
+                    dateTime.getDayOfMonth(),
+                    dateTime.getHour(),
+                    0);
+            case HOURS -> LocalDateTime.of(
+                    dateTime.getYear(),
+                    dateTime.getMonthValue(),
+                    dateTime.getDayOfMonth(),
+                    0,
+                    0);
+            case DAYS -> LocalDateTime.of(
+                    dateTime.getYear(),
+                    dateTime.getMonthValue(),
+                    1,
+                    0,
+                    0);
+            case WEEKS, MONTHS -> LocalDateTime.of(
+                    dateTime.getYear(),
+                    1,
+                    1,
+                    0,
+                    0);
+            case YEARS -> LocalDateTime.of(
+                    1970,
+                    1,
+                    1,
+                    0,
+                    0);
+        };
     }
 }

--- a/unreleased_changes/20230920_092623_738__3757.md
+++ b/unreleased_changes/20230920_092623_738__3757.md
@@ -1,0 +1,19 @@
+* Issue **#3757** : Fix NPE in rule processing.
+
+
+```sh
+# ONLY the top line will be included as a change entry in the CHANGELOG.
+# The entry should be in GitHub flavour markdown and should be written on a SINGLE
+# line with no hard breaks. You can have multiple change files for a single GitHub issue.
+# The  entry should be written in the imperative mood, i.e. 'Fix nasty bug' rather than
+# 'Fixed nasty bug'.
+#
+# Examples of acceptable entries are:
+#
+#
+# * Issue **123** : Fix bug with an associated GitHub issue in this repository
+#
+# * Issue **namespace/other-repo#456** : Fix bug with an associated GitHub issue in another repository
+#
+# * Fix bug with no associated GitHub issue.
+```

--- a/unreleased_changes/20230920_092623_738__3757.md
+++ b/unreleased_changes/20230920_092623_738__3757.md
@@ -1,4 +1,4 @@
-* Issue **#3757** : Fix NPE in rule processing.
+* Issue **#3757** : Fix NPE in rule processing. Fix NPE with `getLastEventTime()`.
 
 
 ```sh

